src/transpose_z_to_y.f90

```diff
@@ -476,4 +429,182 @@ subroutine mem_merge_zy_complex(in, n1, n2, n3, out, iproc, dist, decomp)
 
    end subroutine mem_merge_zy_complex
 
+!!! Utility subroutines
+    
+   subroutine rearrange_sendbuf_real(src, wk1, decomp)
+
+      real(mytype), dimension(:, :, :), intent(IN) :: src
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk1
+#if defined(_GPU)
+      attributes(device) :: wk1
+
+      integer :: istat
+#endif
+
+      integer :: s1, s2, s3
+
+      s1 = SIZE(src, 1)
+      s2 = SIZE(src, 2)
+      s3 = SIZE(src, 3)
+
+#ifdef EVEN
+      if (.not. decomp%even) then
+         call mem_split_zy_real(src, s1, s2, s3, wk1, dims(2), &
+                                decomp%z2dist, decomp)
+      end if
+#else
+      ! note the src array is suitable to be a send buffer
+      ! so no split operation needed
+
+#if defined(_GPU)
+      !$acc host_data use_device(src)
+      istat = cudaMemcpy(wk1, src, s1 * s2 * s3, cudaMemcpyDeviceToDevice)
+      !$acc end host_data
+      if (istat /= 0) call decomp_2d_abort(__FILE__, __LINE__, istat, "cudaMemcpy2D")
+#else
+      associate(foo => wk1, bar => decomp)
+      end associate
+#endif
+
+#endif
+   end subroutine rearrange_sendbuf_real
+
+   subroutine define_recvbuf_real(src, decomp, wk1, wk2)
+
+      implicit none
+
+      real(mytype), dimension(:, :, :), intent(IN) :: src
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk1, wk2
+#if defined(_GPU)
+      attributes(device) :: wk1, wk2
+#endif
+     
+#ifdef EVEN
+      call define_recvbuf_real_even(src, decomp, wk1, wk2)
+#else
+
+#if defined(_GPU)
+      call define_recvbuf_real_gpu(decomp, wk1, wk2)
+#else
+      associate (wk => wk1)
+      end associate
+      call define_recvbuf_real_std(src, decomp, wk2)
+#endif
+#endif
+
+   end subroutine define_recvbuf_real
+
+   subroutine rearrange_recvbuf_real(src, dst, decomp, wk2)
+
+      implicit none
+     
+      real(mytype), dimension(:, :, :), intent(IN) :: src
+      real(mytype), dimension(:, :, :), intent(OUT) :: dst
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk2
+#if defined(_GPU)
+      attributes(device) :: wk2
+#endif
+
+      integer :: s1, s2, s3, d1, d2, d3
+
+      s1 = SIZE(src, 1)
+      s2 = SIZE(src, 2)
+      s3 = SIZE(src, 3)
+      d1 = SIZE(dst, 1)
+      d2 = SIZE(dst, 2)
+      d3 = SIZE(dst, 3)
+
+      call mem_merge_zy_real(wk2, d1, d2, d3, dst, dims(2), &
+           decomp%y2dist, decomp)
+      
+   end subroutine rearrange_recvbuf_real
+
+!!!! Option-specific utility subroutines
+   
+#ifndef EVEN
+   
+#ifndef _GPU
+
+   subroutine define_recvbuf_real_std(src, decomp, wk2)
+
+      implicit none
+
+      real(mytype), dimension(:, :, :), intent(IN) :: src
```

Comment:
I am not sure we need to keep name of the variables from the calling code. It might be easier to read with `src` / `dst` or `wk1` / `wk2`. However it could also be misleading...